File: name-score/name-score-computation/src/main/java/com/shrader/namescore/scoring/strategy/ThirdNameScoreStrategy.java
package com.shrader.namescore.scoring.strategy;

import com.shrader.namescore.scoring.NameScoreStrategy;
import org.springframework.stereotype.Component;

import java.util.Comparator;
import java.util.List;
import java.util.stream.IntStream;


@Component
public class ThirdNameScoreStrategy implements NameScoreStrategy {
    private static final char BASE = 'A';

    @Override
    public int score(List<String> names) {
        names.sort(Comparator.comparing(Object::toString));

        int score = IntStream.range(0, names.size()).parallel()
                .map(i -> scoreName(names.get(i), i + 1))
                .sum();

        return score;
    }

    private int scoreName(String name, int multiplier) {
        int result = 0;

        for (int i = 0; i < name.length(); i++) {
            int ord = name.charAt(i);
            result += ord - BASE + 1;
        }

        return result * multiplier;
    }
}

File: name-score/name-score-computation/src/main/kotlin/com/shrader/namescore/scoring/strategy/ThirdNameScoreStrategy.kt
package com.shrader.namescore.scoring.strategy

import com.shrader.namescore.scoring.NameScoreStrategy
import org.springframework.stereotype.Component
import java.util.stream.IntStream


@Component
class ThirdNameScoreStrategy : NameScoreStrategy {

    companion object {
        private const val BASE = 'A'.toInt()
    }

    override fun score(names: List<String>): Int {
        if(names.isEmpty()) {
            return 0
        }

        val sortedNames = names.sortedWith(String.CASE_INSENSITIVE_ORDER)

        return IntStream.range(0, sortedNames.size).parallel()
                .map { i -> scoreName(sortedNames[i], i + 1) }
                .sum()
    }

    private fun scoreName(name: String, multiplier: Int): Int {
        var result = 0

        for (character in name) {
            val value = character.toInt()
            result += value - BASE + 1
        }

        return result * multiplier
    }
}
